polkadot/node/network/availability-recovery/src/task/strategy/mod.rs

```diff
@@ -0,0 +1,1534 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Recovery strategies.
+
+mod chunks;
+mod full;
+mod systematic;
+
+pub use self::{
+	chunks::{FetchChunks, FetchChunksParams},
+	full::{FetchFull, FetchFullParams},
+	systematic::{FetchSystematicChunks, FetchSystematicChunksParams},
+};
+use crate::{
+	futures_undead::FuturesUndead, ErasureTask, PostRecoveryCheck, RecoveryParams, LOG_TARGET,
+};
+
+use futures::{channel::oneshot, SinkExt};
+use parity_scale_codec::Decode;
+use polkadot_erasure_coding::branch_hash;
+#[cfg(not(test))]
+use polkadot_node_network_protocol::request_response::CHUNK_REQUEST_TIMEOUT;
+use polkadot_node_network_protocol::request_response::{
+	self as req_res, outgoing::RequestError, OutgoingRequest, Recipient, Requests,
+};
+use polkadot_node_primitives::{AvailableData, ErasureChunk};
+use polkadot_node_subsystem::{
+	messages::{AvailabilityStoreMessage, NetworkBridgeTxMessage},
+	overseer, RecoveryError,
+};
+use polkadot_primitives::{AuthorityDiscoveryId, BlakeTwo256, ChunkIndex, HashT, ValidatorIndex};
+use sc_network::{IfDisconnected, OutboundFailure, RequestFailure};
+use std::{
+	collections::{BTreeMap, HashMap, VecDeque},
+	time::Duration,
+};
+
+// How many parallel chunk fetching requests should be running at once.
+const N_PARALLEL: usize = 50;
+
+/// Time after which we consider a request to have failed
+///
+/// and we should try more peers. Note in theory the request times out at the network level,
+/// measurements have shown, that in practice requests might actually take longer to fail in
+/// certain occasions. (The very least, authority discovery is not part of the timeout.)
+///
+/// For the time being this value is the same as the timeout on the networking layer, but as this
+/// timeout is more soft than the networking one, it might make sense to pick different values as
+/// well.
+#[cfg(not(test))]
+const TIMEOUT_START_NEW_REQUESTS: Duration = CHUNK_REQUEST_TIMEOUT;
+#[cfg(test)]
+const TIMEOUT_START_NEW_REQUESTS: Duration = Duration::from_millis(100);
+
+/// The maximum number of times systematic chunk recovery will try making a request for a given
+/// (validator,chunk) pair, if the error was not fatal. Added so that we don't get stuck in an
+/// infinite retry loop.
+pub const SYSTEMATIC_CHUNKS_REQ_RETRY_LIMIT: u32 = 2;
+/// The maximum number of times regular chunk recovery will try making a request for a given
+/// (validator,chunk) pair, if the error was not fatal. Added so that we don't get stuck in an
+/// infinite retry loop.
+pub const REGULAR_CHUNKS_REQ_RETRY_LIMIT: u32 = 5;
```

Comment:
There are 1/3 systematic chunks and 3/3 regular chunks (any chunk is a regular chunk). A systematic chunk is viewed as unique, in the sense that it cannot be replaced by any other chunk. (We need this exact chunk for systematic recovery to work).

So if validator A should hold chunk 1 and the network request fails, only try requesting it once more from validator A. If the second request also fails, don't spend more time requesting it from validator A, go ahead and try from a backer or fall back to regular chunk recovery.

The point of having a smaller retry limit for systematic is that we don't want to spend a lot of time only requesting 1/3 of chunks if we cannot retrieve all of them quickly. We're better off trying regular recovery which has a higher chance of succeeding if a couple of validators don't send their chunks (this enables us to do more requests in parallel). At the same time, regular chunks have a higher retry rate because if regular chunk recovery fails, there are no other strategies to try - the candidate is unavailable.